diff --git a/original/BlockPistonExtension.java b/modified/BlockPistonExtension.java
index 04b68ef4..8a1ae401 100644
--- a/original/BlockPistonExtension.java
+++ b/modified/BlockPistonExtension.java
@@ -16,6 +16,7 @@ public class BlockPistonExtension extends Block {
     public void remove(World world, int i, int j, int k) {
         super.remove(world, i, j, k);
         int l = world.getData(i, j, k);
+        if (l > 5 || l < 0) return; // CraftBukkit - fixed a piston AIOOBE issue.
         int i1 = PistonBlockTextures.a[b(l)];
 
         i += PistonBlockTextures.b[i1];
@@ -138,6 +139,7 @@ public class BlockPistonExtension extends Block {
 
     public void doPhysics(World world, int i, int j, int k, int l) {
         int i1 = b(world.getData(i, j, k));
+        if (i1 > 5 || i1 < 0) return; // CraftBukkit - fixed a piston AIOOBE issue.
         int j1 = world.getTypeId(i - PistonBlockTextures.b[i1], j - PistonBlockTextures.c[i1], k - PistonBlockTextures.d[i1]);
 
         if (j1 != Block.PISTON.id && j1 != Block.PISTON_STICKY.id) {
