diff --git a/original/BlockPressurePlate.java b/modified/BlockPressurePlate.java
index f8e9b215..f9daa00f 100644
--- a/original/BlockPressurePlate.java
+++ b/modified/BlockPressurePlate.java
@@ -3,6 +3,12 @@ package net.minecraft.server;
 import java.util.List;
 import java.util.Random;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.BlockRedstoneEvent;
+import org.bukkit.event.entity.EntityInteractEvent;
+// CraftBukkit end
+
 public class BlockPressurePlate extends Block {
 
     private EnumMobType a;
@@ -89,6 +95,38 @@ public class BlockPressurePlate extends Block {
             flag1 = true;
         }
 
+        // CraftBukkit start - Interact Pressure Plate
+        org.bukkit.World bworld = world.getWorld();
+        org.bukkit.plugin.PluginManager manager = world.getServer().getPluginManager();
+
+        if (flag != flag1) {
+            if (flag1) {
+                for (Object object: list) {
+                    if (object != null) {
+                        org.bukkit.event.Cancellable cancellable;
+
+                        if (object instanceof EntityHuman) {
+                            cancellable = CraftEventFactory.callPlayerInteractEvent((EntityHuman) object, org.bukkit.event.block.Action.PHYSICAL, i, j, k, -1, null);
+                        } else if (object instanceof Entity) {
+                            cancellable = new EntityInteractEvent(((Entity) object).getBukkitEntity(), bworld.getBlockAt(i, j, k));
+                            manager.callEvent((EntityInteractEvent) cancellable);
+                        } else {
+                            continue;
+                        }
+                        if (cancellable.isCancelled()) {
+                            return;
+                        }
+                    }
+                }
+            }
+
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bworld.getBlockAt(i, j, k), flag ? 1 : 0, flag1 ? 1 : 0);
+            manager.callEvent(eventRedstone);
+
+            flag1 = eventRedstone.getNewCurrent() > 0;
+        }
+        // CraftBukkit end
+
         if (flag1 && !flag) {
             world.setData(i, j, k, 1);
             world.applyPhysics(i, j, k, this.id);
