diff --git a/original/BlockSoil.java b/modified/BlockSoil.java
index 2b8068ed..6e700ad5 100644
--- a/original/BlockSoil.java
+++ b/modified/BlockSoil.java
@@ -2,6 +2,11 @@ package net.minecraft.server;
 
 import java.util.Random;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
+// CraftBukkit end
+
 public class BlockSoil extends Block {
 
     protected BlockSoil(int i) {
@@ -46,6 +51,20 @@ public class BlockSoil extends Block {
 
     public void b(World world, int i, int j, int k, Entity entity) {
         if (world.random.nextInt(4) == 0) {
+            // CraftBukkit start - Interact Soil
+            org.bukkit.event.Cancellable cancellable;
+            if (entity instanceof EntityHuman) {
+                cancellable = CraftEventFactory.callPlayerInteractEvent((EntityHuman) entity, org.bukkit.event.block.Action.PHYSICAL, i, j, k, -1, null);
+            } else {
+                cancellable = new EntityInteractEvent(entity.getBukkitEntity(), world.getWorld().getBlockAt(i, j, k));
+                world.getServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+            }
+
+            if (cancellable.isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
+
             world.setTypeId(i, j, k, Block.DIRT.id);
         }
     }
