diff --git a/original/BlockFurnace.java b/modified/BlockFurnace.java
index d06d4384..ab9f6bd9 100644
--- a/original/BlockFurnace.java
+++ b/modified/BlockFurnace.java
@@ -69,6 +69,7 @@ public class BlockFurnace extends BlockContainer {
     public static void a(boolean flag, World world, int i, int j, int k) {
         int l = world.getData(i, j, k);
         TileEntity tileentity = world.getTileEntity(i, j, k);
+        if (tileentity == null) return; // CraftBukkit
 
         c = true;
         if (flag) {
@@ -111,6 +112,7 @@ public class BlockFurnace extends BlockContainer {
         if (!c) {
             TileEntityFurnace tileentityfurnace = (TileEntityFurnace) world.getTileEntity(i, j, k);
 
+            if (tileentityfurnace == null) return; // CraftBukkit
             for (int l = 0; l < tileentityfurnace.getSize(); ++l) {
                 ItemStack itemstack = tileentityfurnace.getItem(l);
 
