diff --git a/original/BlockBloodStone.java b/modified/BlockBloodStone.java
index 526b48a9..b69fbf58 100644
--- a/original/BlockBloodStone.java
+++ b/modified/BlockBloodStone.java
@@ -1,8 +1,22 @@
 package net.minecraft.server;
 
+import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
+
 public class BlockBloodStone extends Block {
 
     public BlockBloodStone(int i, int j) {
         super(i, j, Material.STONE);
     }
+
+    // CraftBukkit start
+    public void doPhysics(World world, int i, int j, int k, int l) {
+        if (net.minecraft.server.Block.byId[l] != null && net.minecraft.server.Block.byId[l].isPowerSource()) {
+            org.bukkit.block.Block block = world.getWorld().getBlockAt(i, j, k);
+            int power = block.getBlockPower();
+
+            BlockRedstoneEvent event = new BlockRedstoneEvent(block, power, power);
+            world.getServer().getPluginManager().callEvent(event);
+        }
+    }
+    // CraftBukkit end
 }
