diff --git a/original/BlockCactus.java b/modified/BlockCactus.java
index 1cbcb8b6..d1800e63 100644
--- a/original/BlockCactus.java
+++ b/modified/BlockCactus.java
@@ -1,5 +1,10 @@
 package net.minecraft.server;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityDamageByBlockEvent;
+import org.bukkit.event.entity.EntityDamageEvent;
+// CraftBukkit end
+
 import java.util.Random;
 
 public class BlockCactus extends Block {
@@ -76,6 +81,21 @@ public class BlockCactus extends Block {
     }
 
     public void a(World world, int i, int j, int k, Entity entity) {
+        // CraftBukkit start - ENTITY_DAMAGEBY_BLOCK event
+        if (entity instanceof EntityLiving) {
+            org.bukkit.block.Block damager = world.getWorld().getBlockAt(i, j, k);
+            org.bukkit.entity.Entity damagee = (entity == null) ? null : entity.getBukkitEntity();
+
+            EntityDamageByBlockEvent event = new EntityDamageByBlockEvent(damager, damagee, EntityDamageEvent.DamageCause.CONTACT, 1);
+            world.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                entity.damageEntity((Entity) null, event.getDamage());
+            }
+            return;
+        }
+        // CraftBukkit end
+
         entity.damageEntity((Entity) null, 1);
     }
 }
