diff --git a/original/EntityWolf.java b/modified/EntityWolf.java
index eb52c936..61ae30a8 100644
--- a/original/EntityWolf.java
+++ b/modified/EntityWolf.java
@@ -3,6 +3,15 @@ package net.minecraft.server;
 import java.util.Iterator;
 import java.util.List;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason;
+import org.bukkit.event.entity.EntityTargetEvent;
+// CraftBukkit end
+
 public class EntityWolf extends EntityAnimal {
 
     private boolean a = false;
@@ -97,7 +106,17 @@ public class EntityWolf extends EntityAnimal {
             List list = this.world.a(EntitySheep.class, AxisAlignedBB.b(this.locX, this.locY, this.locZ, this.locX + 1.0D, this.locY + 1.0D, this.locZ + 1.0D).b(16.0D, 4.0D, 16.0D));
 
             if (!list.isEmpty()) {
-                this.setTarget((Entity) list.get(this.world.random.nextInt(list.size())));
+                // CraftBukkit start
+                Entity entity = (Entity) list.get(this.world.random.nextInt(list.size()));
+                org.bukkit.entity.Entity bukkitTarget = entity == null ? null : entity.getBukkitEntity();
+
+                EntityTargetEvent event = new EntityTargetEvent(this.getBukkitEntity(), bukkitTarget, EntityTargetEvent.TargetReason.RANDOM_TARGET);
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (!event.isCancelled() || event.getTarget() != null ) {
+                    this.setTarget(entity);
+                }
+                // CraftBukkit end
             }
         }
 
@@ -228,8 +247,21 @@ public class EntityWolf extends EntityAnimal {
         } else {
             if (!this.isTamed() && !this.isAngry()) {
                 if (entity instanceof EntityHuman) {
-                    this.setAngry(true);
-                    this.target = (Entity) entity;
+                    // CraftBukkit start
+                    org.bukkit.entity.Entity bukkitTarget = entity == null ? null : entity.getBukkitEntity();
+
+                    EntityTargetEvent event = new EntityTargetEvent(this.getBukkitEntity(), bukkitTarget, EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY);
+                    this.world.getServer().getPluginManager().callEvent(event);
+
+                    if (!event.isCancelled()) {
+                        if (event.getTarget() == null) {
+                            this.target = null;
+                        } else {
+                            this.setAngry(true);
+                            this.target = ((CraftEntity) event.getTarget()).getHandle();
+                        }
+                    }
+                    // CraftBukkit end
                 }
 
                 if (entity instanceof EntityArrow && ((EntityArrow) entity).shooter != null) {
@@ -245,10 +277,23 @@ public class EntityWolf extends EntityAnimal {
                         EntityWolf entitywolf = (EntityWolf) entity1;
 
                         if (!entitywolf.isTamed() && entitywolf.target == null) {
-                            entitywolf.target = (Entity) entity;
-                            if (entity instanceof EntityHuman) {
-                                entitywolf.setAngry(true);
+                            // CraftBukkit start
+                            org.bukkit.entity.Entity bukkitTarget = entity == null ? null : entity.getBukkitEntity();
+
+                            EntityTargetEvent event = new EntityTargetEvent(this.getBukkitEntity(), bukkitTarget, EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY);
+                            this.world.getServer().getPluginManager().callEvent(event);
+
+                            if (!event.isCancelled()) {
+                                if (event.getTarget() == null) {
+                                    this.target = null;
+                                } else {
+                                    entitywolf.target = (Entity) entity;
+                                    if (entity instanceof EntityHuman) {
+                                        entitywolf.setAngry(true);
+                                    }
+                                }
                             }
+                            // CraftBukkit end
                         }
                     }
                 }
@@ -286,6 +331,17 @@ public class EntityWolf extends EntityAnimal {
             if (this.isTamed()) {
                 b0 = 4;
             }
+            // CraftBukkit start
+            org.bukkit.entity.Entity damager = this.getBukkitEntity();
+            org.bukkit.entity.Entity damagee = entity == null ? null : entity.getBukkitEntity();
+
+            EntityDamageByEntityEvent event = new EntityDamageByEntityEvent(damager, damagee, EntityDamageEvent.DamageCause.ENTITY_ATTACK, b0);
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
 
             entity.damageEntity(this, b0);
         }
@@ -302,7 +358,9 @@ public class EntityWolf extends EntityAnimal {
                 }
 
                 if (!this.world.isStatic) {
-                    if (this.random.nextInt(3) == 0) {
+                    // CraftBukkit - added event call and isCancelled check.
+                    if (this.random.nextInt(3) == 0 && !CraftEventFactory.callEntityTameEvent(this, entityhuman).isCancelled()) {
+                        // CraftBukkit end
                         this.setTamed(true);
                         this.setPathEntity((PathEntity) null);
                         this.setSitting(true);
@@ -328,7 +386,7 @@ public class EntityWolf extends EntityAnimal {
                         entityhuman.inventory.setItem(entityhuman.inventory.itemInHandIndex, (ItemStack) null);
                     }
 
-                    this.b(((ItemFood) Item.PORK).k());
+                    this.b(((ItemFood) Item.PORK).k(), RegainReason.EATING);
                     return true;
                 }
             }
