diff --git a/original/EntityFish.java b/modified/EntityFish.java
index e793a9ea..ff63566d 100644
--- a/original/EntityFish.java
+++ b/modified/EntityFish.java
@@ -2,6 +2,13 @@ package net.minecraft.server;
 
 import java.util.List;
 
+// CraftBukkit start
+import org.bukkit.entity.Projectile;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.event.player.PlayerFishEvent;
+// CraftBukkit end
+
 public class EntityFish extends Entity {
 
     private int d = -1;
@@ -182,7 +189,28 @@ public class EntityFish extends Entity {
 
             if (movingobjectposition != null) {
                 if (movingobjectposition.entity != null) {
-                    if (movingobjectposition.entity.damageEntity(this.owner, 0)) {
+                    // CraftBukkit start
+                    // TODO add EntityDamagedByProjectileEvent : fishing hook?
+                    boolean stick;
+                    if (movingobjectposition.entity instanceof EntityLiving) {
+                        org.bukkit.entity.Entity damagee = movingobjectposition.entity.getBukkitEntity();
+                        Projectile projectile = (Projectile) this.getBukkitEntity();
+
+                        // TODO @see EntityArrow#162
+                        EntityDamageByEntityEvent event = new EntityDamageByEntityEvent(projectile, damagee, EntityDamageEvent.DamageCause.PROJECTILE, 0);
+                        this.world.getServer().getPluginManager().callEvent(event);
+
+                        if (event.isCancelled()) {
+                            stick = !projectile.doesBounce();
+                        } else {
+                            // this function returns if the fish should stick in or not, i.e. !bounce
+                            stick = movingobjectposition.entity.damageEntity(this, event.getDamage());
+                        }
+                    } else {
+                        stick = movingobjectposition.entity.damageEntity(this.owner, 0);
+                    }
+                    if (!stick) {
+                        // CraftBukkit end
                         this.c = movingobjectposition.entity;
                     }
                 } else {
@@ -309,6 +337,16 @@ public class EntityFish extends Entity {
         byte b0 = 0;
 
         if (this.c != null) {
+            // CraftBukkit start
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) this.owner.getBukkitEntity(), this.c.getBukkitEntity(), PlayerFishEvent.State.CAUGHT_ENTITY);
+            this.world.getServer().getPluginManager().callEvent(playerFishEvent);
+
+            if (playerFishEvent.isCancelled()) {
+                this.die();
+                this.owner.hookedFish = null;
+                return 0;
+            }
+            // CraftBukkit end
             double d0 = this.owner.locX - this.locX;
             double d1 = this.owner.locY - this.locY;
             double d2 = this.owner.locZ - this.locZ;
@@ -321,6 +359,16 @@ public class EntityFish extends Entity {
             b0 = 3;
         } else if (this.k > 0) {
             EntityItem entityitem = new EntityItem(this.world, this.locX, this.locY, this.locZ, new ItemStack(Item.RAW_FISH));
+            // CraftBukkit start
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) this.owner.getBukkitEntity(), entityitem.getBukkitEntity(), PlayerFishEvent.State.CAUGHT_FISH);
+            this.world.getServer().getPluginManager().callEvent(playerFishEvent);
+
+            if (playerFishEvent.isCancelled()) {
+                this.die();
+                this.owner.hookedFish = null;
+                return 0;
+            }
+            // CraftBukkit end
             double d5 = this.owner.locX - this.locX;
             double d6 = this.owner.locY - this.locY;
             double d7 = this.owner.locZ - this.locZ;
@@ -336,9 +384,25 @@ public class EntityFish extends Entity {
         }
 
         if (this.h) {
+            // CraftBukkit start
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) this.owner.getBukkitEntity(), null, PlayerFishEvent.State.IN_GROUND);
+            this.world.getServer().getPluginManager().callEvent(playerFishEvent);
+
+            if (playerFishEvent.isCancelled()) {
+                this.die();
+                this.owner.hookedFish = null;
+                return 0;
+            }
+            // CraftBukkit end
             b0 = 2;
         }
 
+        // CraftBukkit start
+        if (b0 == 0) {
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) this.owner.getBukkitEntity(), null, PlayerFishEvent.State.FAILED_ATTEMPT);
+            this.world.getServer().getPluginManager().callEvent(playerFishEvent);
+        }
+        // CraftBukkit end
         this.die();
         this.owner.hookedFish = null;
         return b0;
