diff --git a/original/EntityItem.java b/modified/EntityItem.java
index 7770e304..d8558b67 100644
--- a/original/EntityItem.java
+++ b/modified/EntityItem.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.bukkit.event.player.PlayerPickupItemEvent; // CraftBukkit
+
 public class EntityItem extends Entity {
 
     public ItemStack itemStack;
@@ -8,6 +10,7 @@ public class EntityItem extends Entity {
     public int pickupDelay;
     private int f = 5;
     public float d = (float) (Math.random() * 3.141592653589793D * 2.0D);
+    private int lastTick = (int) (System.currentTimeMillis() / 50); // CraftBukkit
 
     public EntityItem(World world, double d0, double d1, double d2, ItemStack itemstack) {
         super(world);
@@ -15,6 +18,11 @@ public class EntityItem extends Entity {
         this.height = this.width / 2.0F;
         this.setPosition(d0, d1, d2);
         this.itemStack = itemstack;
+        // CraftBukkit start - infinite item fix
+        if (this.itemStack.count <= -1) {
+            this.itemStack.count = 1;
+        }
+        // CraftBukkit end
         this.yaw = (float) (Math.random() * 360.0D);
         this.motX = (double) ((float) (Math.random() * 0.20000000298023224D - 0.10000000149011612D));
         this.motY = 0.20000000298023224D;
@@ -35,9 +43,11 @@ public class EntityItem extends Entity {
 
     public void m_() {
         super.m_();
-        if (this.pickupDelay > 0) {
-            --this.pickupDelay;
-        }
+        // CraftBukkit start
+        int currentTick = (int) (System.currentTimeMillis() / 50);
+        this.pickupDelay -= (currentTick - this.lastTick);
+        this.lastTick = currentTick;
+        // CraftBukkit end
 
         this.lastX = this.locX;
         this.lastY = this.locY;
@@ -113,6 +123,24 @@ public class EntityItem extends Entity {
         if (!this.world.isStatic) {
             int i = this.itemStack.count;
 
+            // CraftBukkit start
+            int canHold = entityhuman.inventory.canHold(this.itemStack);
+            int remaining = this.itemStack.count - canHold;
+            if (this.pickupDelay <= 0 && canHold > 0) {
+                this.itemStack.count = canHold;
+                PlayerPickupItemEvent event = new PlayerPickupItemEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
+                this.world.getServer().getPluginManager().callEvent(event);
+                this.itemStack.count = canHold + remaining;
+
+                if (event.isCancelled()) {
+                    return;
+                }
+
+                // Possibly < 0; fix here so we do not have to modify code below
+                this.pickupDelay = 0;
+            }
+            // CraftBukkit end
+
             if (this.pickupDelay == 0 && entityhuman.inventory.pickup(this.itemStack)) {
                 if (this.itemStack.id == Block.LOG.id) {
                     entityhuman.a((Statistic) AchievementList.g);
