diff --git a/original/ItemFishingRod.java b/modified/ItemFishingRod.java
index 6dd083d3..9bcf2e2f 100644
--- a/original/ItemFishingRod.java
+++ b/modified/ItemFishingRod.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.bukkit.event.player.PlayerFishEvent; // CraftBukkit
+
 public class ItemFishingRod extends Item {
 
     public ItemFishingRod(int i) {
@@ -15,6 +17,14 @@ public class ItemFishingRod extends Item {
             itemstack.damage(i, entityhuman);
             entityhuman.w();
         } else {
+            // CraftBukkit start
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), null,PlayerFishEvent.State.FISHING);
+            world.getServer().getPluginManager().callEvent(playerFishEvent);
+
+            if (playerFishEvent.isCancelled()) {
+                return itemstack;
+            }
+            // CraftBukkit end
             world.makeSound(entityhuman, "random.bow", 0.5F, 0.4F / (b.nextFloat() * 0.4F + 0.8F));
             if (!world.isStatic) {
                 world.addEntity(new EntityFish(world, entityhuman));
