diff --git a/original/ItemHoe.java b/modified/ItemHoe.java
index 40686d69..08854516 100644
--- a/original/ItemHoe.java
+++ b/modified/ItemHoe.java
@@ -1,5 +1,11 @@
 package net.minecraft.server;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.block.CraftBlockState;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.BlockPlaceEvent;
+// CraftBukkit end
+
 public class ItemHoe extends Item {
 
     public ItemHoe(int i, EnumToolMaterial enumtoolmaterial) {
@@ -21,7 +27,19 @@ public class ItemHoe extends Item {
             if (world.isStatic) {
                 return true;
             } else {
+                CraftBlockState blockState = CraftBlockState.getBlockState(world, i, j, k); // CraftBukkit
+
                 world.setTypeId(i, j, k, block.id);
+
+                // CraftBukkit start - Hoes - blockface -1 for 'SELF'
+                BlockPlaceEvent event = CraftEventFactory.callBlockPlaceEvent(world, entityhuman, blockState, i, j, k, block);
+
+                if (event.isCancelled() || !event.canBuild()) {
+                    event.getBlockPlaced().setTypeId(blockState.getTypeId());
+                    return false;
+                }
+                // CraftBukkit end
+
                 itemstack.damage(1, entityhuman);
                 return true;
             }
