diff --git a/original/PropertyManager.java b/modified/PropertyManager.java
index c8bf9e6a..6a186b54 100644
--- a/original/PropertyManager.java
+++ b/modified/PropertyManager.java
@@ -10,7 +10,7 @@ import java.util.logging.Logger;
 public class PropertyManager {
 
     public static Logger a = Logger.getLogger("Minecraft");
-    private Properties properties = new Properties();
+    public Properties properties = new Properties(); // CraftBukkit - priv to pub
     private File c;
 
     public PropertyManager(File file1) {
@@ -28,6 +28,24 @@ public class PropertyManager {
         }
     }
 
+    // CraftBukkit start
+    private joptsimple.OptionSet options = null;
+
+    public PropertyManager(final joptsimple.OptionSet options) {
+        this((File) options.valueOf("config"));
+
+        this.options = options;
+    }
+
+    private <T> T getOverride(String name, T value) {
+        if ((this.options != null) && (this.options.has(name))) {
+            return (T) this.options.valueOf(name);
+        }
+
+        return value;
+    }
+    // CraftBukkit end
+
     public void a() {
         a.log(Level.INFO, "Generating new properties file");
         this.savePropertiesFile();
@@ -44,17 +62,19 @@ public class PropertyManager {
 
     public String getString(String s, String s1) {
         if (!this.properties.containsKey(s)) {
+            s1 = this.getOverride(s, s1); // CraftBukkit
             this.properties.setProperty(s, s1);
             this.savePropertiesFile();
         }
 
-        return this.properties.getProperty(s, s1);
+        return this.getOverride(s, this.properties.getProperty(s, s1)); // CraftBukkit
     }
 
     public int getInt(String s, int i) {
         try {
-            return Integer.parseInt(this.getString(s, "" + i));
+            return this.getOverride(s, Integer.parseInt(this.getString(s, "" + i))); // CraftBukkit
         } catch (Exception exception) {
+            i = this.getOverride(s, i); // CraftBukkit
             this.properties.setProperty(s, "" + i);
             return i;
         }
@@ -62,14 +82,16 @@ public class PropertyManager {
 
     public boolean getBoolean(String s, boolean flag) {
         try {
-            return Boolean.parseBoolean(this.getString(s, "" + flag));
+            return this.getOverride(s, Boolean.parseBoolean(this.getString(s, "" + flag))); // CraftBukkit
         } catch (Exception exception) {
+            flag = this.getOverride(s, flag); // CraftBukkit
             this.properties.setProperty(s, "" + flag);
             return flag;
         }
     }
 
     public void b(String s, boolean flag) {
+        flag = this.getOverride(s, flag); // CraftBukkit
         this.properties.setProperty(s, "" + flag);
         this.savePropertiesFile();
     }
