diff --git a/original/WorldGenTrees.java b/modified/WorldGenTrees.java
index 79d5ff0e..f8fd4710 100644
--- a/original/WorldGenTrees.java
+++ b/modified/WorldGenTrees.java
@@ -2,11 +2,23 @@ package net.minecraft.server;
 
 import java.util.Random;
 
+import org.bukkit.BlockChangeDelegate; // CraftBukkit
+
 public class WorldGenTrees extends WorldGenerator {
 
     public WorldGenTrees() {}
 
     public boolean a(World world, Random random, int i, int j, int k) {
+        // CraftBukkit start
+        // sk: The idea is to have (our) WorldServer implement
+        // BlockChangeDelegate and then we can implicitly cast World to
+        // WorldServer (a safe cast, AFAIK) and no code will be broken. This
+        // then allows plugins to catch manually-invoked generation events
+        return this.generate((BlockChangeDelegate) world, random, i, j, k);
+    }
+
+    public boolean generate(BlockChangeDelegate world, Random random, int i, int j, int k) {
+        // CraftBukkit end
         int l = random.nextInt(3) + 4;
         boolean flag = true;
 
