diff --git a/original/WorldManager.java b/modified/WorldManager.java
index f4fde503..8c6c9923 100644
--- a/original/WorldManager.java
+++ b/modified/WorldManager.java
@@ -3,7 +3,7 @@ package net.minecraft.server;
 public class WorldManager implements IWorldAccess {
 
     private MinecraftServer server;
-    private WorldServer world;
+    public WorldServer world; // CraftBukkit - private -> public
 
     public WorldManager(MinecraftServer minecraftserver, WorldServer worldserver) {
         this.server = minecraftserver;
@@ -13,11 +13,11 @@ public class WorldManager implements IWorldAccess {
     public void a(String s, double d0, double d1, double d2, double d3, double d4, double d5) {}
 
     public void a(Entity entity) {
-        this.server.getTracker(this.world.worldProvider.dimension).track(entity);
+        this.server.getTracker(this.world.dimension).track(entity); // CraftBukkit
     }
 
     public void b(Entity entity) {
-        this.server.getTracker(this.world.worldProvider.dimension).untrackEntity(entity);
+        this.server.getTracker(this.world.dimension).untrackEntity(entity); // CraftBukkit
     }
 
     public void a(String s, double d0, double d1, double d2, float f, float f1) {}
@@ -27,7 +27,7 @@ public class WorldManager implements IWorldAccess {
     public void a() {}
 
     public void a(int i, int j, int k) {
-        this.server.serverConfigurationManager.flagDirty(i, j, k, this.world.worldProvider.dimension);
+        this.server.serverConfigurationManager.flagDirty(i, j, k, this.world.dimension); // CraftBukkit
     }
 
     public void a(String s, int i, int j, int k) {}
@@ -37,6 +37,6 @@ public class WorldManager implements IWorldAccess {
     }
 
     public void a(EntityHuman entityhuman, int i, int j, int k, int l, int i1) {
-        this.server.serverConfigurationManager.sendPacketNearby(entityhuman, (double) j, (double) k, (double) l, 64.0D, this.world.worldProvider.dimension, new Packet61(i, j, k, l, i1));
+        this.server.serverConfigurationManager.sendPacketNearby(entityhuman, (double) j, (double) k, (double) l, 64.0D, this.world.dimension, new Packet61(i, j, k, l, i1)); // CraftBukkit
     }
 }
